subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/BuildOperationValidator.groovy

```diff
@@ -0,0 +1,95 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.fixtures
+
+import groovy.json.JsonSlurper
+import org.gradle.integtests.fixtures.executer.GradleExecuter
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.util.TextUtil
+
+
+class BuildOperationValidator {
+    private final GradleExecuter executer
+    private final TestFile projectDir
+    private final TestFile operationsDir
+    private Map results
+
+    BuildOperationValidator(GradleExecuter executer, TestFile projectDir) {
+        this.executer = executer
+        this.projectDir = projectDir
+        this.operationsDir = projectDir.file("operations")
+    }
+
+    BuildOperationValidator build() {
+        def initScript = projectDir.file("buildOpValidator.gradle")
+        initScript << """
+            import org.gradle.internal.progress.BuildOperationService
+            import org.gradle.internal.progress.BuildOperationListener
+            import org.gradle.internal.progress.BuildOperationInternal
+            import org.gradle.internal.progress.OperationStartEvent
+            import org.gradle.internal.progress.OperationResult
+
+            def operations = [:]
+            def operationListener = new BuildOperationListener() {
+                
+                void started(BuildOperationInternal buildOperation, OperationStartEvent startEvent) {
+                    operations[buildOperation.id] = [
+                        displayName: "\${buildOperation.displayName}",
+                        parentId: "\${buildOperation.parentId}",
+                        name: "\${buildOperation.name}",
+                        startTime: startEvent.startTime
+                    ]
+                }
+
+                void finished(BuildOperationInternal buildOperation, OperationResult finishEvent) {
+                    if (operations.containsKey(buildOperation.id)) {
+                        operations[buildOperation.id].endTime = finishEvent.endTime
+                        if (finishEvent.failure != null) {
+                            operations[buildOperation.id].failure = finishEvent.failure.message
+                        }
+                    }
+                }
+            }
+
+            gradle.services.get(BuildOperationService).addListener(operationListener)
+
+            gradle.buildFinished {
+                def operationsDir = new File("${TextUtil.normaliseFileSeparators(operationsDir.absolutePath)}")
+                operationsDir.mkdirs()
+                def jsonFile = new File(operationsDir, "operations.json")
+                def json = new groovy.json.JsonBuilder()
+                json.operations(operations)
+                jsonFile << json.toPrettyString()
+            }
+        """
+        executer.usingInitScript(initScript)
+        return this
+    }
+
+    Map getOperations() {
```

Comment:
Sure, would be nice.
`ProgressEvents` lays out operations in a tree so we can query/assert that tree structure
The only real difference is in how the events are captured.
They could share a model, queries and assertions helpers.